### PR TITLE
Always use relative internal links in rich text fields

### PR DESCRIPTION
**Pull request by chosak**
This commit restores behavior that was removed in #2872.

When we insert internal Wagtail page links into rich text fields, we need them to always render as relative (not absolute) so as to maintain expected behavior across sharing/production sites. This new handler makes sure all internal Wagtail page links are always rendered as relative.

See [comments in `v1.wagtail_hooks.RelativePageLinkHandler`](https://github.com/cfpb/cfgov-refresh/compare/better-link-handler?expand=1#diff-44507cc860654b805baac659b58d2c53R118) for more details.

## Additions

- New `v1.wagtail_hooks.RelativePageLinkHandler` that forces relative internal links in Wagtail rich text fields.

## Testing

- New unit tests cover the new handler class; to manually test locally, [create a `BrowsePage`](http://localhost:8000/admin/pages/add/v1/browsepage/3/), add a `FullWidthText` organism, insert a "Content" `RichTextBlock`, then insert an internal link to another Wagtail page. You'll see that links render as relative when previewing or viewing the page even if multiple Wagtail `Site`s are defined.

## Notes

- As always, [I should have listened better](https://github.com/cfpb/cfgov-refresh/pull/2872#issuecomment-297458526) to @Scotchester.

## Checklist

* [X] PR has an informative and human-readable title
* [X] Changes are limited to a single goal (no scope creep)
* [X] Code can be automatically merged (no conflicts)
* [X] Passes all existing automated tests
* [X] New functions include new tests
* [X] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Project documentation has been updated
* [X] Reviewers requested with the [Assignee tool](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) :arrow_right:

### CODE REVIEW

**Comment by willbarton** [comment]
What are the implications for #2873?

File: cfgov/v1/tests/test_wagtail_hooks.py [hunk 1]
```diff
@@ -118,3 +117,49 @@ def test_calls_child_block_is_submitted(self, mock_getstreamfields, mock_hasattr
...
+            '<a>',
+        )
+
+    def test_page_not_in_site_root_returns_none_link(self):
+        self.assertEqual(
+            RelativePageLinkHandler.expand_db_attributes(
+                attrs={'id': 1},
+                for_editor=False
+            ),
+            '<a href="None">',
```
Comment (willbarton) [inline cfgov/v1/tests/test_wagtail_hooks.py]: Would it be better to behave like a nonexistent page if the page is outside the site root?

File: cfgov/v1/wagtail_hooks.py [hunk 2]
```diff
@@ -110,3 +113,57 @@ def register_django_admin_menu_item():
...
+                editor_attrs = ''
+
+            page_url = page.specific.url
+
+            if page_url:
+                page_url = urlsplit(page_url).path
+
+            return '<a %shref="%s">' % (editor_attrs, escape(page_url))
+        except Page.DoesNotExist:
+            return "<a>"
```
Comment (willbarton) [inline cfgov/v1/wagtail_hooks.py]: Is it only ever when the page doesn't exist in the site root that `page.specific.url` might evaluate to `False` here?

**Comment by chosak** [comment]
@willbarton good question, thanks for bringing that up. I think #2873 can probably be closed as, for now, this does the same thing. If/when this gets merged in, I'll close that one.

File: cfgov/v1/tests/test_wagtail_hooks.py (hunk 1 above)
Comment (chosak) [inline cfgov/v1/tests/test_wagtail_hooks.py]: I kept this behavior [consistent with how Wagtail does it](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/rich_text.py#L50). To clarify, it's not "the site root", it's "any site root" - meaning this only appears if a link is to a page that isn't associated with any `Site` at all (wouldn't even have an absolute URL).

**Comment by Scotchester** [comment]
@chosak Hmm... judging by the changed code, #2873 looks like a more elegant solution. Are you sure you wouldn't rather go that route?

File: cfgov/v1/wagtail_hooks.py (hunk 2 above)
Comment (chosak) [inline cfgov/v1/wagtail_hooks.py]: Yes, if a page isn't under a site root that's been exposed via a `Site`.

File: cfgov/v1/tests/test_wagtail_hooks.py (hunk 1 above)
Comment (willbarton) [inline cfgov/v1/tests/test_wagtail_hooks.py]: Ah, alright. Consistency with Wagtail is good... I just wonder when that would happen.

**Comment by chosak** [comment]
@Scotchester I didn't listen well enough to you last time so I have to be extra thoughtful in my reply now :smile:

I think I prefer this implementation as being somewhat more explicit. This PR is limited to "when rendering page links in rich text fields, always use relative URLs". The motivation for all of these changes has generally been based on how the HTML is generated, so tying this to the rendering code seems more straightforward to me.

Changing `CFGOVPage.url` has more wide-ranging impact: for example when working in the Python shell or in other unit tests, all of a sudden your page object acts differently. It's also slightly more hidden in the possibly unlikely case that we fork page classes in future.

For both approaches, the reason we need this is due to our non-standard use of a separate domain for content editing. In the long run it would be better to move away from that setup. Keeping this customization in the realm of the rendering code might also make it easier to remove if/when we get to that point.

File: cfgov/v1/tests/test_wagtail_hooks.py (hunk 1 above)
Comment (chosak) [inline cfgov/v1/tests/test_wagtail_hooks.py]: A recent example that happened to us was when a live page contained a link to a "deleted" page that had been moved to the Trash. As the Trash isn't under a site root, it turned the live page's link into `<a href="None">`.

Comment (chosak) [inline cfgov/v1/tests/test_wagtail_hooks.py]: A recent example that happened to us was when a live page contained a link to a "deleted" page that had been moved to the Trash. As the Trash isn't under a site root, it turned the live page's link into `<a href="None">`.

**Comment by Scotchester** [comment]
Great answer :+1:

> the reason we need this is due to our non-standard use of a separate domain for content editing

For content _editing_ or for _viewing draft_ content? Do you have an idea for how we'd be able to review draft content without a different domain?

**Comment by chosak** [comment]
I meant _editing_ in particular.

Keeping things somewhat vague as this is public, one approach might be to move editing/admin access to the public domain that we serve pages from while keeping the existing internal sharing domain only for viewing of draft content. This would require replacing current `Site`-based feature flags with some other mechanism (e.g. query string) as well as secure limited exposure of `/admin` on that public domain (cc @willbarton).

If we went this route, Wagtail would only have a single `Site`, and thus all `Page.url` calls would be relative everywhere without needing custom handlers.

**Comment by Scotchester** [comment]
Just so I'm clear: The rendered result should be `<a href="none"> for any page that isn't published and in the current site, correct?

This is what my testing bears out.

File: cfgov/v1/wagtail_hooks.py [hunk 3]
```diff
@@ -3,9 +3,12 @@
 from django.core.exceptions import PermissionDenied
 from django.conf import settings
 from django.core.urlresolvers import reverse
-from django.utils.html import format_html_join
+from django.utils.html import escape, format_html_join
+from urlparse import urlsplit
```
Comment (Scotchester) [inline cfgov/v1/wagtail_hooks.py]: Maybe put blank lines above and below the `urlparse` import to separate the groups?

File: cfgov/v1/wagtail_hooks.py [hunk 4]
```diff
@@ -110,3 +113,57 @@ def register_django_admin_menu_item():
...
+
+    This special page link handler makes it so that any internal Wagtail page
+    links inserted into rich text fields, for example:
+
+        <a linktype="page" id="123">foo</a>
+
+    always get replaced with a relative URL to that page, never an absolute
+    one, like this:
+
+        <a href="/path/to/page">foo</a>
```
Comment (Scotchester) [inline cfgov/v1/wagtail_hooks.py]: The above explanation could maybe be a little clearer about the fact that the first example is the database representation, and the second example is the final rendering to the client.

Also, the phrasing makes it unclear to me if the second example is a relative URL or an absolute URL. Consider including an example of each?

File: cfgov/v1/wagtail_hooks.py [hunk 5]
```diff
@@ -110,3 +113,57 @@ def register_django_admin_menu_item():
...
+
+    In standard Wagtail behavior, this may be an absolute URL if an
```
Comment (Scotchester) [inline cfgov/v1/wagtail_hooks.py]: `may` ➡️ `will`

File: cfgov/v1/wagtail_hooks.py [hunk 6]
```diff
@@ -110,3 +113,57 @@ def register_django_admin_menu_item():
...
+    installation has multiple Wagtail Sites. In our current custom usage we
+    have multiple Wagtail Sites (one for production, one for staging) that
+    share the same root page. So forcing the use of relative URLs would work
+    fine and allow for easier navigation within a single domain.
+
+    This will explicitly break things if users ever wanted to host some
+    additional site that doesn't share the same root page.
+
+    This code is modified from `wagtail.wagtailcore.rich_text.PageLinkHandler`.
+
```
Comment (Scotchester) [inline cfgov/v1/wagtail_hooks.py]: Trivial: You could remove this blank line to match our typical docstring formatting

**Comment by chosak** [comment]
@Scotchester that's correct. Thanks for the thorough review; I've applied your excellent style suggestions in 3202686.

**Pull request merged**